### PR TITLE
SCF-Session-4

### PR DESCRIPTION
**Pull request by meta-ishita-mathur**

### CODE REVIEW

File: SCF-Session-4/mirror/MirrorTest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	public void FourthMaxTest() 
+	{
+		int maxCount = mirror.maxMirror(new int[]{1, 4, 5, 3, 5, 4, 1}, 7);
+		assertEquals(7, maxCount);
+	}
+	
+	@Test(expected = AssertionError.class)
+	public void testException()
+	{
+		int maxCount = mirror.maxMirror(new int[]{}, 0);
```
Comment (sandeep-sky) [inline SCF-Session-4/mirror/MirrorTest.java]: Variable `maxCount` is not used after declaration. Why it is declared?

File: SCF-Session-4/split/SplitArrayTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	public void testNegative()
+	{
+		int indexSplit = splitArray.split(new int[]{2, 1, 1, 2, 1}, 5);
+		assertEquals(-1, indexSplit);
+	}
+	
+	@org.junit.Test(expected = AssertionError.class)
+	public void testException()
+	{
+		int indexSplit = splitArray.split(new int[]{}, 0);
```
Comment (sandeep-sky) [inline SCF-Session-4/split/SplitArrayTest.java]: Variable `indexSplit` is not used after declaration. Why it is declared?

File: SCF-Session-4/clumps/ArrayClumps.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public class ArrayClumps 
+{
+	/**
+	 * This method returns the number of clumps in an array.
+	 * @param inputArray, should contain a series of 2 or more adjacent elements of the same value
+	 * @param arrayLength > 0
+	 * @return the count of clumps in the array.
+	 */
+	public int countClumps(int[] inputArray, int arrayLength)
```
Comment (sandeep-sky) [inline SCF-Session-4/clumps/ArrayClumps.java]: Why second parameter `arrayLength` is needed? Length can be calculated through `inputArray.length`.

File: SCF-Session-4/clumps/ArrayClumps.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	{
+		int occur = 0, count = 0;
+		
+		if (arrayLength == 0)
+			throw new AssertionError("Array is empty");
+		
+		for(int i = 0; i < arrayLength; i++)
+		{
+			for(int j = i + 1; j < arrayLength; j++)
+			{
```
Comment (sandeep-sky) [inline SCF-Session-4/clumps/ArrayClumps.java]: Nested loop is not really needed to resolve this. Try to resolve with single loop.

File: SCF-Session-4/fixXY/ArrayFixXY.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	 * 
+	 * @param inputArray, should contain equal number of X and Y 
+	 * & two adjacent X values should not be there & X should not be at the last position. 
+	 * @param arrayLength > 0
+	 * @param x, this is the value that should not move.
+	 * @param y, this is the value that should come after x.
+	 * @return
+	 */
+	public int[] fixXY(int[] inputArray, int arrayLength, int x, int y)
+	{
```
Comment (sandeep-sky) [inline SCF-Session-4/fixXY/ArrayFixXY.java]: Parameter `arrayLength` is not needed. `inputArray.length` can be used to calculate array length.

File: SCF-Session-4/fixXY/ArrayFixXY.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+			if (inputArray[j] == inputArray[j + 1] && inputArray[j] == x)
+				throw new AssertionError("Two adjacent X values");
+		}
+		
+		for(int k = 0; k <= arrayLength - 1; k++)
+		{
+			if (k == arrayLength - 1 && inputArray[k] == x)
+				throw new AssertionError("X is at last position");
+		}
+
```
Comment (sandeep-sky) [inline SCF-Session-4/fixXY/ArrayFixXY.java]: Above two `for loops` are not needed to resolve this. Use the single `for loop` to resolve.

File: SCF-Session-4/fixXY/ArrayFixXY.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	public int[] swap(int index, int[] newArray, int arrayLength, int X, int Y)
+	{
+		int tempSwap, findSwap = 0;
+		
+		for(int i = 0; i < arrayLength; i++)
+		{
+			if(newArray[i] == Y)
+			{
+				if(i == 0 || newArray[i-1] != X)
+				{
```
Comment (sandeep-sky) [inline SCF-Session-4/fixXY/ArrayFixXY.java]: Why nested if? Single statement can be used if there is only one block of code to execute.

File: SCF-Session-4/split/SplitArray.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+		for(i = 1; i <= arrayLength - 1; i++)
+		{
+			firstPart = 0;
+			lastPart = 0;
+			
+			for(j = i - 1; j >= 0; j--)
+				firstPart += inputArray[j];
+			
+			for(k = i; k < arrayLength; k++)
+				lastPart += inputArray[k];
```
Comment (sandeep-sky) [inline SCF-Session-4/split/SplitArray.java]: Why these nested loops?

File: SCF-Session-4/mirror/MirrorTest.java (hunk 1 above)
Comment (meta-ishita-mathur) [reply SCF-Session-4/mirror/MirrorTest.java]: done

File: SCF-Session-4/split/SplitArrayTest.java (hunk 2 above)
Comment (meta-ishita-mathur) [reply SCF-Session-4/split/SplitArrayTest.java]: done

File: SCF-Session-4/fixXY/ArrayFixXY.java (hunk 5 above)
Comment (meta-ishita-mathur) [reply SCF-Session-4/fixXY/ArrayFixXY.java]: done

File: SCF-Session-4/fixXY/ArrayFixXY.java (hunk 6 above)
Comment (meta-ishita-mathur) [reply SCF-Session-4/fixXY/ArrayFixXY.java]: done

File: SCF-Session-4/fixXY/ArrayFixXY.java (hunk 7 above)
Comment (meta-ishita-mathur) [reply SCF-Session-4/fixXY/ArrayFixXY.java]: done

File: SCF-Session-4/split/SplitArray.java (hunk 8 above)
Comment (meta-ishita-mathur) [reply SCF-Session-4/split/SplitArray.java]: done

File: SCF-Session-4/mirror/MaxMirror.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		if (arrayLength == 0)
+			throw new AssertionError("Array is empty");
+	   
+	    for (int i = 0; i < arrayLength; i++)
+	    {
+	        tempCount = 1;
+	        mirrorCount = i;
+
+	        for (int j = arrayLength - 1; j >= 0; j--)
+	        {
```
[the remark this answers is not in the document]
Comment (meta-ishita-mathur) [reply SCF-Session-4/mirror/MaxMirror.java]: done

**Pull request merged**